keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+
+Because the first step succeeds, also the second step succeeds. So we don’t need error handling during defaulting. The third step will catch wrong types of the used defaults.
+
+## Formal Proposal – following pruning option 1
+
+The examples in the motivational section show that the semantics of full OpenAPI validation schemata are not trivial in respect to pruning and defaulting. To simplify the algorithms and to enforce "sane" schemata which allow to split schema and value validation, we propose to derive a skeleton schema from the full user-given OpenAPI validation schema
+* which does not contain value validations
+* which is complete enough for pruning and defaulting.
+
+**Remark:** we have two options to define and implement pruning and defaulting:
+1. directly on the full OpenAPI validation schema with two custom algorithms doing parallel recursion over the schema and the input object.
+2. via an intermediate representation (the skeleton schema) and using go-openapi pruning and defaulting. Both algorithms are ten-liners based on the go-openapi/validate output. With the skeleton schema the go-openapi pruning algorithm coincides with our pruning option 1.
+
+Both routes lead to the same algorithm: the intermediate representation of the skeleton schema makes merging of OpenAPI validation schema constraints explicit, while the custom algorithms would hide that in its recursion code.
+
+Moreover, the main reason for the intermediate schema: the custom algorithm would have to replicate a lot of the validation logic of go-openapi, e.g. the semantics of `properties`, `additionalPropoerties`, `patternProperties` and the same for items of an array. With route 2 we get all of this for free. 
+
+### Example 7: 
+
+> Assume
+>
+>```json
+>{
+>    "patternProperties": {
+>	    "^foo-.*": { "properties": {"a":{}}, "c": {"default": 1} },
+>		"^bar-.*": { "properties": {"b":{}}, "c": {"default": 2} }
+>	}
+>}
+>```
+>
+>An object `{"foo-42": {"a": 1, "b": 2}, "bar-42": {"a": 1", "b": 2}}` will be pruned to `{"foo-42": {"a": 1}, "bar-42": {"b": 2}}` and defaulted to `{"foo-42": {"a": 1, "c": 1}, "bar-42": {"b": 2, "c": 2}}`.
+>
+>Note that the default and the pruning only depends on the object keys, not on values.
+
+### Definition: skeleton schema
+
+> For a given OpenAPI validation schema `s` the skeleton schema `skel(s)` is derived by 
+> 1. applying `skel` to all elements of `.allOf`, `.anyOf`, `.oneOf` and `.not` giving `s_1, …, s_n`,
+> then dropping all fields from `s` other than 
+> * `type`, 
+> * `items`,
+> * `additionalItems`,
+> * `properties`,
+> * `patternProperties`,
+> * `additionalProperties`,
+> * `default`,
+> giving `drop(s)`,
+> 2. then merging `s_i` into it’s containing object.
+>
+>    I.e.: `skel(s) := merge(drop(s), s_1, …, s_n)` with
+>
+>```
+>merge(x_1, …, x_n) := { 
+>   "type": t if all x_i agree on t as type, undefined otherwise
```

Comment:
**Mehdy Bohlool:** does type matter for pruning/defaulting?

**Stefan Schimanski:** no, see discussion about type and format below. We can leave out type+format completely.

**Stefan Schimanski:** added comment about that.